connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala

```diff
@@ -57,28 +69,105 @@ private[connect] object ErrorUtils extends Logging {
+        .newBuilder()
+        .setMessage(error.getMessage)
+        .addAllErrorTypeHierarchy(ErrorUtils.allClasses(error.getClass).map(_.getName).asJava)
+
+      if (isTruncated) {
+        builder.setMessage(SparkConnectService.extractErrorMessage(error, MAX_MESSAGE_SIZE))
+      } else {
+        if (stackTraceInMessage) {
+          if (serverStackTraceEnabled || pySparkJVMStackTraceEnabled) {
+            builder.setMessage(
+              s"${error.getMessage}\n\n" +
+                s"JVM stacktrace:\n${ExceptionUtils.getStackTrace(error)}")
+          }
+        } else {
+          if (serverStackTraceEnabled) {
+            builder.addAllStackTrace(
+              error.getStackTrace
+                .map { stackTraceElement =>
+                  FetchErrorDetailsResponse.StackTraceElement
+                    .newBuilder()
+                    .setDeclaringClass(stackTraceElement.getClassName)
+                    .setMethodName(stackTraceElement.getMethodName)
+                    .setFileName(stackTraceElement.getFileName)
+                    .setLineNumber(stackTraceElement.getLineNumber)
+                    .build()
+                }
+                .toIterable
+                .asJava)
+          }
+        }
+      }
+
+      lastErrorOpt.foreach(builder.setCause(_))
+      lastErrorOpt = Some(builder.build())
+    }
+
+    val responseBuilder = FetchErrorDetailsResponse.newBuilder()
+
+    lastErrorOpt.foreach(responseBuilder.setExceptionInfo(_))
+
+    responseBuilder.build()
+  }
+
+  private def buildStatusFromThrowable(
+      st: Throwable,
+      enrichErrorEnabled: Boolean,
+      userId: String,
+      sessionId: String): RPCStatus = {
     val errorInfo = ErrorInfo
       .newBuilder()
       .setReason(st.getClass.getName)
       .setDomain("org.apache.spark")
-      .putMetadata(
-        "classes",
-        JsonMethods.compact(JsonMethods.render(allClasses(st.getClass).map(_.getName))))
-
-    lazy val stackTrace = Option(ExceptionUtils.getStackTrace(st))
-    val withStackTrace = if (stackTraceEnabled && stackTrace.nonEmpty) {
-      val maxSize = SparkEnv.get.conf.get(Connect.CONNECT_JVM_STACK_TRACE_MAX_SIZE)
-      errorInfo.putMetadata("stackTrace", StringUtils.abbreviate(stackTrace.get, maxSize))
-    } else {
-      errorInfo
+      .putMetadata("classes", serializeClasses(st))
+
+    if (enrichErrorEnabled) {
+      val errorId = UUID.randomUUID().toString
+
+      // The errorId of the exception.
+      errorInfo.putMetadata("errorId", errorId)
+
+      SparkConnectService
+        .getOrCreateIsolatedSession(userId, sessionId)
+        .errorIdToError
+        .put(errorId, st)
     }
 
     RPCStatus
       .newBuilder()
       .setCode(RPCCode.INTERNAL_VALUE)
-      .addDetails(ProtoAny.pack(withStackTrace.build()))
-      .setMessage(SparkConnectService.extractErrorMessage(st))
+      .addAllDetails(
+        Seq(errorInfo.build(), throwableToFetchErrorDetailsResponse(st))
```

Comment:
Why do you need to send an abbreviated FetchErrorDetailsResponse here? We did say that the user has to explicitly, fetch the exception.

Reply:
IIRC the agreed upon behavior was that we keep everything as is, but we simply add an error ID that the client can use to fetch the full error stack trace.

Reply:
In case the additional RPC failed, fallback to an abbreviated FetchErrorDetailsResponse is consistent with the existing control. Otherwise, we may have to throw an Exception or fallback to an exception without cause support.